src/ssl/support.cc

```diff
@@ -213,14 +244,24 @@ int Ssl::matchX509CommonNames(X509 *peer_cert, void *check_data, int (*check_fun
         int numalts = sk_GENERAL_NAME_num(altnames);
         for (int i = 0; i < numalts; ++i) {
             const GENERAL_NAME *check = sk_GENERAL_NAME_value(altnames, i);
-            if (check->type != GEN_DNS) {
-                continue;
-            }
-            ASN1_STRING *cn_data = check->d.dNSName;
-
-            if ( (*check_func)(check_data, cn_data) == 0) {
-                sk_GENERAL_NAME_pop_free(altnames, GENERAL_NAME_free);
-                return 1;
+            switch(check->type) {
+                case GEN_DNS:
+                    // If the type is GEN_DNS, call check_func with the dNSName data
+                    if ( (*check_func)(check_data, check->d.dNSName) == 0) {
+                        sk_GENERAL_NAME_pop_free(altnames, GENERAL_NAME_free);
+                        return 1;
+                    }
+                    break;
+                case GEN_IPADD:
+                    debugs(83, 4, "Check type is GEN_IPADD, verifying...");
+                    // If it's an IP address, attempt to compare it with the check_data
+                    if (compare_ip_addresses(check_data, check->d.iPAddress) == 0) {
+                        sk_GENERAL_NAME_pop_free(altnames, GENERAL_NAME_free);
+                        return 1;
```

Comment:
This GEN_IPADD `case` addition feels like a Ssl::matchX509CommonNames() API violation to me. The function promises to (and did) outsource "yes, there is a match" conclusion to check_func. This PR-added code ignores that promise.

Ssl::matchX509CommonNames() API is used by several code paths. I suspect that many (if not all) of those paths would benefit from being able to handle GEN_IPADD entries.

I suspect that we have to adjust Ssl::matchX509CommonNames() API to replace check_func function with a C++ class object that has two methods: matchDomainName() and matchIp().

Reply:
So, I've added 2 new commits. The most recent adds a `VerifyAddress` class. However, so that I don't have to mess with the calling semantics for matchX509CommonNames in other locations, I've retained the API to expect a `check_func` (or equivalent) but modified the body of the original `check_func` to use the new `VerifyAddress::match` method.

Reply:
FWIW - I found it more simple to retain the `switch` and then ensure the correct `ASN1_STRING` is passed to `match`. So we're still outsourcing the check/conclusion to check_func but being purposeful about which data we send to it for comparison.